.azure-pipelines/release-cli.yaml

```diff
@@ -56,658 +52,680 @@ parameters:
-            displayName: DotNet publish
-
-          - template: templates/setup-powershell-environment.yml
-
-          - task: PowerShell@2
-            inputs:
-              targetType: inline
-              script: |
-                New-Item $(outputDir) -ItemType Directory -Force
-                echo "Test file" > $(outputDir)/mgc
-                echo "Test file2" > $(outputDir)/mgc.txt
-              verbosePreference: '$(OUTPUT_PREFERENCE)'
-              debugPreference: '$(OUTPUT_PREFERENCE)'
-              informationPreference: '$(OUTPUT_PREFERENCE)'
-            condition: and(succeeded(), eq('${{ parameters.simulate }}', 'true'))
-            displayName: Simulate publish
-
-          - task: PowerShell@2
-            inputs:
-              pwsh: true
-              targetType: inline
-              script: |
-                . $(powershellScriptsDir)/BuildTools.ps1
-                $archive = Compress-BuildOutput -OutputDir '$(outputDir)-$(rid)' -Source '$(outputDir)/*' -FileNameTemplate '$(fileNameTemplate)' -BranchOrTagName '$(branchOrTagName)' -RuntimeIdentifier '$(rid)' -PackageType '$(packageType)' -TarCompression $(compressionProgram) -Cleanup
-                $hasArchive = Test-Path $archive -PathType Leaf
-                Write-Host "##vso[task.setvariable variable=HAS_ARTIFACTS]$hasArchive"
-              verbosePreference: '$(OUTPUT_PREFERENCE)'
-              debugPreference: '$(OUTPUT_PREFERENCE)'
-              informationPreference: '$(OUTPUT_PREFERENCE)'
-            # Only create artifacts if we're building a tag, building main, or building a PR targeting main
-            condition: and(succeeded(), or(eq('${{ parameters.forceSignOutput }}', 'true'), startsWith(variables['Build.SourceBranch'], 'refs/tags/v'), eq(variables['Build.SourceBranch'], 'refs/heads/main'), eq(variables['System.PullRequest.TargetBranch'], 'main')))
-            displayName: Compress published output
-
-          - task: PublishPipelineArtifact@1
-            # Only publish artifacts if they exist
-            condition: and(succeeded(), eq(variables['HAS_ARTIFACTS'], 'True'))
-            inputs:
-              artifact: build-output-$(rid)
-              path: $(outputDir)-$(rid)
-
-    # Format takes a long time. keep it in its own stage
-  - stage: checkFormat
-    displayName: Check formatting
-    condition: false
-    dependsOn: []
-    jobs:
-      - job: checkFormat
+            - task: DotNetCoreCLI@2
+              displayName: Run tests
+              enabled: true
+              inputs:
+                command: test
+                workingDirectory: $(Build.SourcesDirectory)
+                arguments: --no-restore
+
+      - stage: build
+        displayName: Build CLI
+        dependsOn: [test]
+        jobs:
+          - job: build
+            displayName: Build and publish
+            strategy:
+              matrix:
+                'Windows-x64':
```

Comment:
I struggled with matrices changing the agent when trying to do that for kiota. Is this actually switching the agent OS somehow? or are they all running on linux at this point?

Reply:
This section runs on linux. The matrix doesn't change the agent OS. That has to be done in the `pool` options. I've done that in the ESRP section, but I haven't tested it.
https://github.com/microsoftgraph/msgraph-cli/blob/eb44476d34b16da3c05ead4127bfbe3d5c3ff2b3/.azure-pipelines/release-cli.yaml#L394-L397

Reply:
thanks for the context. The challenge I have been facing is that 1es template expects jobs to be known in advance so it can inject all the compliance stuff. So matrices are an issue to start with.
Then, if you use the loop solution they provide, a mismatch between the base agent and the job agent triggers an error from the 1es template. I have reached out multiple times to the team without success.
If you manage to make it work for the CLI, let me know, I'll use your learnings to finish the 1es template migration for kiota.

Reply:
Bad news. It didn't work. The MacOS images failed to run :-(

Reply:
Can we get a run where it exercises all stages up until *upload*?

Reply:
I think I figured it out...

Reply:
I can see you've moved to the each syntax like I've tried, but I'm not sure how you're switching agent for each of the jobs? could you provide some more context please?

Reply:
I'm not switching agents. The jobs get created with the agent already selected. The agents and pools are configured in the parameters.
https://github.com/microsoftgraph/msgraph-cli/blob/95b5db959e95f72988138d69954968e10487ecec/.azure-pipelines/release-cli.yaml#L163-L195

I've also added links to the information source.

Reply:
ooh nice! let me know once it's tested, I'll try to adapt the kiota one. Out of curiosity, where did you get that solution? I couldn't find anything comprehensive in their docs

Reply:
I searched the channel on teams and saw another team that had asked about matrix pipelines. I then took 1ES's suggestion to use parameters from the comments on the question.